### PR TITLE
Migration utilities

### PR DESCRIPTION
**Pull request by rco-odoo**
Those are a bunch of functions to help writing migration scripts.

### CODE REVIEW

File: odoo/migration/migration.py [hunk 1]
```diff
@@ -0,0 +1,2383 @@
...
+    """
+    cr.execute(q, (table,))
+    return cr.fetchall()
+
+
+def delete_unused(cr, table, xmlids, set_noupdate=True):
+    """
+    Delete all records in the provided list not being referenced in any FKEY.
+
+    Note that the xmlids th
```
Comment (Yajo) [inline odoo/migration/migration.py]: :thinking: :question:

**Comment by pedrobaeza** [comment]
It seems this is competence of `openupgradelib` :stuck_out_tongue_closed_eyes: 

Out of jokes, good move!

File: odoo/migration/migration.py (hunk 1 above)
Comment (Yajo) [inline odoo/migration/migration.py]: DRY. This already exists: https://github.com/odoo/odoo/blob/d6ee2abb987016c28e24f471f4b3d413af84c185/odoo/sql_db.py#L406-L418

Comment (Yajo) [inline odoo/migration/migration.py]: :thinking: :question:

**Comment by pedrobaeza** [comment]
Are these methods object of review or it's only a disclosure of them?

**Comment by dbo-odoo** [comment]
Note that this PR is certainly *not* ready to be reviewed.

At this stage, I just took a big utility file we used across all versions of odoo/openerp (from 7.sthg up to master) for migrations; this code had several layers of compatibility shims (python 2/3,  framework idiosyncracies across versions) and helper functions that now exist in the framework but that had to be included in this lib for older version (since it was meant to be version agnostic, which will no longer be the case with this lib that will evolve with future versions).

At this stage, I've just taken this utility, removed some dust, added some docstrings to understand what we're dealing with and removed *a few things*. @rco-odoo being more competent than me, he will take up the work, remove what's redundant, rename a few functions, add consistency to function names etc.

Then again, it's more a disclosure of these functions and an addition as a utility library in the framework than a call for contribution at this stage (I think). Perhaps we can discuss improving upon it and merging some ideas/concepts from openupgrade; I'd very much like to discuss this with @pedrobaeza at Odoo Experience :)

I leave @rco-odoo manage his work as he wishes, I'm just glad this will be available to all and that it'll allow devs of odoo.sh/Odoo Apps & OCA to write migration scripts more easily (or at all, in some cases -_-)

**Comment by pedrobaeza** [comment]
OK, I let @rco-odoo to tell when it's ready for being reviewed (if he wants to), and very glad to discuss this with you @dbo-odoo and who wants to join at Odoo XP (or at least know each other) :smiley:

**Comment by blaggacao** [comment]
It would be really nice getting this out as soon as convenient. I want to use those scripts *now*...

**Comment by rco-odoo** [comment]
Task has been cancelled.

**Comment by pedrobaeza** [comment]
I hope it has been cancelled for being better done later :smiley:

File: odoo/tools/migration.py [hunk 2]
```diff
@@ -0,0 +1,2383 @@
...
+                   JOIN dups d ON (r.{c2} = ANY(d.others))
+                 EXCEPT
+                 SELECT r.{c1}, r.{c2}
+                   FROM {rel} r
+                   JOIN dups d ON (r.{c2} = d.id)
+        """.format(rel=ft, c1=cols[0], c2=fc))
+
+    assert upds         # if not m2m found, there is something wrong...
+
+    updates = ','.join('_upd_%s AS (%s)' % x for x in enumerate(upds))
```
Comment (blaggacao) [inline odoo/tools/migration.py]: What should the first `%s` originally take as a value? - I doubt the whole query... :wink:

Comment (dbo-odoo) [reply odoo/tools/migration.py]: you might want to recheck what enumerate does then; your faith is not required :stuck_out_tongue: 
``` SQL
_upd_0 AS (INSERT INTO ...etc
```

Comment (blaggacao) [reply odoo/tools/migration.py]: Where is this simely banging my hand against the front. Thank you ;-)

**Comment by blaggacao** [comment]
@dbo-odoo There was a Identifier not quoted issue somwhere (I beleive). I bothered to normalize the queries using the `pyscop2.sql` package. For what it's worth: https://gist.github.com/blaggacao/50f5b13f5b489071bec3029ec9a4f65c

Would it be lost investment (due to some information I do not have access to) for me do cover those scripts with some tests?

**Comment by dbo-odoo** [comment]
> Would it be lost investment (due to some information I do not have access to) for me do cover those scripts with some tests?

Yes, it probably would. :disappointed:

File: odoo/tools/migration.py [hunk 3]
```diff
@@ -0,0 +1,2383 @@
...
+
+        By using `uniq_column=lower(name)` and `order=name`
+        you can prioritize tags in CamelCase/UPPERCASE.
+    """
+    table = table_of_model(cr, model)
+    upds = []
+    for ft, fc, _, da in get_fk(cr, table):
+        assert da == 'c'    # should be a ondelete=cascade fk
+        cols = get_columns(cr, ft, ignore=(fc,))[0]
+        assert len(cols) == 1   # it's a m2, should have only 2 columns
```
Comment (blaggacao) [inline odoo/tools/migration.py]: This breaks the following migration:
```python
migration.uniq_tags(cr, "res.partner.category", uniq_column="name", order="id")
```
for 
```python
get_fk(cr, table)
```
returns 
```python
[('res_partner_category', 'parent_id', 'res_partner_category_parent_id_fkey', 'c'), ('res_partner_res_partner_category_rel', 'category_id', 'res_partner_res_partner_category_rel_category_id_fkey', 'c')]

```
cause it's a hierarchical table. 

Note the first element references itself and therefore has more columns than a pure relational table.

Either this was not meant for hierarchy table (though most `tags` ara hierarchies) or this function is flawed.

Which one is the case?

**Comment by blaggacao** [comment]
I did successfully test: https://gist.github.com/blaggacao/50f5b13f5b489071bec3029ec9a4f65c
against: https://github.com/xoe-labs/dodoo-migrator/blob/master/tests/test_odoo_migration.py

Was 3 hours, so I don't mind if it's in vain. For the meantime it serves it's purpose.
:smile:

**Pull request closed without merging**